src/algebra/hom/centroid.lean

```diff
@@ -0,0 +1,289 @@
+@[simp] lemma comp_id (f : centroid_hom α) : f.comp (centroid_hom.id α) = f := ext $ λ a, rfl
+@[simp] lemma id_comp (f : centroid_hom α) : (centroid_hom.id α).comp f = f := ext $ λ a, rfl
+
+lemma cancel_right {g₁ g₂ f : centroid_hom α} (hf : surjective f) :
+  g₁.comp f = g₂.comp f ↔ g₁ = g₂ :=
+⟨λ h, ext $ hf.forall.2 $ fun_like.ext_iff.1 h, congr_arg _⟩
+
+lemma cancel_left {g f₁ f₂ : centroid_hom α} (hg : injective g) : g.comp f₁ = g.comp f₂ ↔ f₁ = f₂ :=
+⟨λ h, ext $ λ a, hg $ by rw [←comp_apply, h, comp_apply], congr_arg _⟩
+
+instance : has_zero (centroid_hom α) :=
+⟨{ map_mul_left' := λ a b, (mul_zero _).symm,
+  map_mul_right' := λ a b, (zero_mul _).symm,
+  ..(0 : α →+ α) }⟩
+
+instance : has_one (centroid_hom α) := ⟨centroid_hom.id α⟩
+
+instance : has_add (centroid_hom α) :=
+⟨λ f g,
+  { map_mul_left' := λ a b, by simp [map_mul_left, mul_add],
+    map_mul_right' := λ a b, by simp [map_mul_right, add_mul],
+    ..(f + g : α →+ α) } ⟩
+
+instance : has_mul (centroid_hom α) := ⟨comp⟩
+
+instance has_scalar_nat : has_smul ℕ (centroid_hom α) :=
+⟨λ n f,
+  { map_mul_left' := λ a b,
+      by { change n • f (a * b) = a * n • f b, rw [map_mul_left f, ←mul_smul_comm] },
+    map_mul_right' := λ a b,
+      by { change n • f (a * b) = n • f a * b, rw [map_mul_right f, ←smul_mul_assoc] },
+    .. (n • f : α →+ α) }⟩
+
+instance has_npow_nat : has_pow (centroid_hom α) ℕ :=
+⟨λ f n,
+{  map_mul_left' := λ a b, begin
+    induction n with n ih,
+    { simp },
+    { rw pow_succ,
+      exact (congr_arg f.to_End ih).trans (f.map_mul_left' _ _) }
+  end,
+  map_mul_right' := λ a b, begin
+    induction n with n ih,
+    { simp },
+    { rw pow_succ,
+      exact (congr_arg f.to_End ih).trans (f.map_mul_right' _ _) }
+  end,
+  ..(f.to_End ^ n : add_monoid.End α) }⟩
+
+@[simp] lemma coe_zero : ⇑(0 : centroid_hom α) = 0 := rfl
+@[simp] lemma coe_one : ⇑(1 : centroid_hom α) = id := rfl
+@[simp] lemma coe_add (f g : centroid_hom α) : ⇑(f + g) = f + g := rfl
+@[simp] lemma coe_mul (f g : centroid_hom α) : ⇑(f * g) = f ∘ g := rfl
+-- Eligible for `dsimp`
+@[simp, nolint simp_nf] lemma coe_nsmul (f : centroid_hom α) (n : ℕ) : ⇑(n • f) = n • f := rfl
+
+@[simp] lemma zero_apply (a : α) : (0 : centroid_hom α) a = 0 := rfl
+@[simp] lemma one_apply (a : α) : (1 : centroid_hom α) a = a := rfl
+@[simp] lemma add_apply (f g : centroid_hom α) (a : α) : (f + g) a = f a + g a := rfl
+@[simp] lemma mul_apply (f g : centroid_hom α) (a : α) : (f * g) a = f (g a) := rfl
+-- Eligible for `dsimp`
+@[simp, nolint simp_nf]
+lemma nsmul_apply (f : centroid_hom α) (n : ℕ) (a : α) : (n • f) a = n • f a := rfl
+
+-- cf.`add_monoid_hom.add_comm_monoid`
+instance : add_comm_monoid (centroid_hom α) :=
+coe_to_add_monoid_hom_injective.add_comm_monoid _ rfl (λ _ _, rfl) (λ _ _, rfl)
+
+instance : add_comm_monoid_with_one (centroid_hom α) :=
+{ ..centroid_hom.add_comm_monoid,
+  ..centroid_hom.has_one }
```

Comment:
I recommend setting `nat_cast` explictly here, so that the proof below is refl. I would guess that looks something like
```suggestion
instance : has_nat_cast (centroid_hom α) :=
{ nat_cast := \lam n, n • 1 }
```

Reply:
```
instance : has_nat_cast (centroid_hom α) :=
{ nat_cast := λ n, n • 1 }
```
seems to work okay. Did you mean that I should use the `refl` tactic in the proof of `instance : add_comm_monoid_with_one (centroid_hom α)` ? It's not immediately obvious to me what that should look like?